Popover: fix arrow placement and design

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Fixes #42820

## What?
<!-- In a few words, what is the PR actually doing? -->

Fix a regression introduced in #40740 about the `Popover`'s arrow placement.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The `Popover`'s arrow has currently a few issues:

- it is not always positioned correctly
- it doesn't render with the same background color and border as the popover's contents
- its size is smaller than it used to be before #40740

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

In order to fix the arrow placement, the following changes were implemented:

- fixed a bug in the logic behind validating the arrow's `x` and `y` coordinates received from `useFloating` (mainly switching from `! Number.isNaN` to `Number.isFinite`)
- changed the approach in rendering the arrow's triangle:
  - from a diamond rotated `45deg`, switched to a proper triangle (coded in SVG)
  - Added (back) support for a border color around the triangle, in line with how the design used to be
  - moved some inline styles to SASS styles, using an internal `$arrow-triangle-base-size` SCSS variable

I've also added two extra temporary commits (which I will remove before merging)
  - one adding a Storybook example to easily check that the arrow's placement is correct for all possible placements
  - another one adding a quick change that makes sure that `ownerDocument` is always defined inside `Popover` (this fix should probably be moved to its own separate PR)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- In Storybook: on your machine, visit the new temporary "Positioning" Storybook example and make sure that the arrow's placement and design looks correct for all possible popover placements
- In the WordPress app: on your machine, visit `/wp-admin/site-editor.php?postType=wp_template` and click the "Add New" button in the top right part of the screen. Make sure that the Popover's arrow placement and design look correct (see [this comment](https://github.com/WordPress/gutenberg/issues/42820#issuecomment-1200612126) for more details)
- Try to replicate the bug reported in [the original issue's description](https://github.com/WordPress/gutenberg/issues/42820#issue-1322276575) and confirm that it can't be replicated anymore

For extra thoroughness, you could revert the `Popover: fix arrow logic, use SVG to render triangle` commit in this PR and verify that the bug is not fixed in the scenarios described above.

## Screenshots or screencast <!-- if applicable -->

| Before | After |
|---|---|
|  ![image](https://user-images.githubusercontent.com/1083581/182248194-04c0dd46-b92b-40f4-ab66-791ab39fed31.png) | ![Screenshot 2022-08-01 at 23 19 57](https://user-images.githubusercontent.com/1083581/182248609-fab94cf9-ddb0-4b7a-97c8-6ec9bb7ec117.png) |


**Before:**

https://user-images.githubusercontent.com/1083581/182247869-27cb9797-e8d3-4cd8-a45e-d43d89f52774.mp4


**After:**

https://user-images.githubusercontent.com/1083581/182247281-7f07e868-6305-40b2-9a8c-40b4f0bb2e0b.mp4

ciampo committed:
Popover: fix arrow logic, use SVG to render triangle

ciampo committed:
CHANGELOG

ciampo committed:
Add RTL support

mirka commented on packages/components/src/popover/index.js:
```diff
@@ -407,22 +428,25 @@ const Popover = (
-					className="components-popover__arrow"
 					ref={ arrowRef }
+					className={ [
+						'components-popover__arrow',
+						`components-popover__arrow--${
+							placementData.split( '-' )[ 0 ]
+						}`,
```
The [CSS coding guidelines](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#css) in this repo are "inspired by BEM" but not exactly BEM, so we want the modifier classes to be named `.is-top`, `.is-right`, etc.

ntsekouras replied:
Super nit: why not just use string interpolation here?

ciampo replied:
> The [CSS coding guidelines](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md#css) in this repo are "inspired by BEM" but not exactly BEM, so we want the modifier classes to be named `.is-top`, `.is-right`, etc.

Done in https://github.com/WordPress/gutenberg/pull/42874/commits/097510867480c0187c5e536d16a0a0f11eecb731


> Super nit: why not just use string interpolation here?

I'm not sure I understand your suggestion, can you elaborate a bit more?

ntsekouras replied:
I was suggesting:
```
className={`components-popover__arrow is-${ placementData.split( '-' )[ 0 ] }`}
```

ciampo replied:
I see, thank you for the eplanation!

I personally tend to use the `[ ...classnames ].join(' ')` syntax out of habit — I find it tidier to read, and it allows me to easily add conditional classnames + filter to exclude them if needed (ie.. `[ ...classnames ].filter(Boolean).join(' ')`)

mirka commented on packages/components/src/popover/style.scss:
```diff
@@ -60,12 +60,64 @@
+	stroke-width: $border-width;
+
+	// Same as popover content
+	&-bg {
+		fill: $white;
+		stroke: $white;
+	}
+
+	&-border {
```
Nit: This might be a personal preference thing, but I prefer full class names to be written out (`.components-popover__triangle-bg`) so they can be searched easily 🥺

ciampo replied:
Done in https://github.com/WordPress/gutenberg/pull/42874/commits/4a00b8dd36f33c0eff09669629e3ba29237db451

mirka commented on packages/components/src/popover/style.scss:
```diff
@@ -60,12 +60,64 @@
-	z-index: -1;
-	.is-alternate & {
-		background: $gray-900;
+	width: var(--wp-popover-arrow-size);
+	height: var(--wp-popover-arrow-size);
```
Nice 💛

mirka reviewed (approved):
Looks great, thanks for the quick fix!

At first I was wondering why the Add New Template dropdown is offset a little bit lower than [before](https://user-images.githubusercontent.com/555336/182060058-c240c365-1f40-4165-8e83-001a5069d410.png) and no longer overlaps the header, but I guess we can consider this a consistency improvement since it's the same offset as other Dropdowns without the arrow. Like the Options dropdown in the editor:

<img width="346" alt="The Options dropdown in the editor sits flush with the header" src="https://user-images.githubusercontent.com/555336/182296641-cfc556f0-ddd1-421f-ae86-b906fcabb88b.png">

ntsekouras commented on packages/components/src/popover/index.js:
```diff
@@ -137,25 +162,26 @@ const Popover = (
 		}
 
-		return document;
+		return documentToReturn ?? document;
```
I guess there are cases that `documentToReturn` is not defined to make this change.. Can you explain where that is and if you know why? Do we miss something in our current handling of `anchorRef`?

ciampo replied:
Sure! I had to add this change because, when I created the temporary Storybook example, I decided to [pass the anchor to the popover via `anchorRef`](https://github.com/WordPress/gutenberg/pull/42874/commits/759d0348398f44c820ab8d50d15275cca1b2ab6b#diff-88eb39a8c17230c5315615af25bf97fe202a13447ab24d5c6cf3cf292ff6ba82R111).

That caused an error because `ownerDocument` was `undefined` — and so I inspected the code:

https://github.com/WordPress/gutenberg/blob/759d0348398f44c820ab8d50d15275cca1b2ab6b/packages/components/src/popover/index.js#L139-L159

Basically, what I believe was happening was:
- `anchorRef` is a "modern style" ref, and therefore it's an object with the `current` prop
- initially, `anchorRef.current` is `null`
- therefore, the check on line `144` (`if ( anchorRef?.current )`) returns `false`
- but the check on line `146` (`if ( anchorRef )`) returns `true`. and therefore `ownerDocument` ends up being `undefined` because `anchorRef` doesn't have an `ownerDocument` property

If you agree, I was planning on opening a separate PR with just this fix once this PR is merged

ciampo replied:
https://github.com/WordPress/gutenberg/pull/42886

ciampo commented:
> At first I was wondering why the Add New Template dropdown is offset a little bit lower than [before](https://user-images.githubusercontent.com/555336/182060058-c240c365-1f40-4165-8e83-001a5069d410.png) and no longer overlaps the header, but I guess we can consider this a consistency improvement since it's the same offset as other Dropdowns without the arrow.

I see both "sides" of the argument, but I agree with your point — this could be seen as a step towards consistency. But even if we changed our mind and wanted to move the popover closer to the "Add new" button, I would suggest that we do it in a separate PR so that we keep this one focused on the arrow fix.

ntsekouras commented on packages/components/src/popover/index.js:
```diff
@@ -407,22 +428,25 @@ const Popover = (
+						top: Number.isFinite( arrowData?.y )
+							? `${ arrowData.y }px`
+							: '',
+						'--wp-popover-arrow-size': '14px',
```
So `14px` is our magic number and I guess it's the one to match our previous design, correct? 

Since this number is not affected by anything and is not configurable, is there a reason to have it here and not in the scss file directly?

ciampo replied:
> So `14px` is our magic number and I guess it's the one to match our previous design, correct?

Short answer, yes!

Long answer: after a few trials, it looks like, in order to match our previous designs:

- in terms of arrow proportions, the height of the triangle needs to be roughly half of its base (which reflects in the path coordinates in the SVG)
- the previous "arrow size" before the popover refactor was `8px`
- therefore, I factored a height for the triangle of `7px`(plus `1px` border)
- therefore, the base of the triangle is 7 * 2 = `14px`

> Since this number is not affected by anything and is not configurable, is there a reason to have it here and not in the scss file directly?

Good question! When I started working on this PR, most of the arrow styles where applied inline and required some form of math in JS (mostly to get the triangle height, knowing the target triangle base). And so, I decided to use a CSS custom prop to share the value between JS and SCSS.

But now that you point it out, there isn't really a need for it — so I switched to a SCSS variable (which follows more the style of the component) in https://github.com/WordPress/gutenberg/pull/42874/commits/de33f60b3ce48ebcc9edc53bab00f02cab86db7a

ciampo committed:
Use modifier classes instead of traditional BEM

ntsekouras reviewed (approved):
Awesome work here Marco! 💯 

I took a brief look at this yesterday and saw the many things missing from arrow handling. I've tested your PR by changing every Dropdown we have to show an arrow and everything seemed great(both in post and site editors).

I've left a few small comments/questions and I'd like to know more especially for [this one](https://github.com/WordPress/gutenberg/pull/42874/files#r935191862). I know you'll have some small cleanup to do, but this looks good to me!

ciampo committed:
Use full classnames for triangle SVG

talldan commented:
FWIW, the Add New Template dropdown probably shouldn't have an arrow. I think it's the only dropdown in the codebase that has one, so that could be seen as an inconsistency.

ciampo commented:
> FWIW, the Add New Template dropdown probably shouldn't have an arrow. I think it's the only dropdown in the codebase that has one, so that could be seen as an inconsistency.

I'm not an expert when it comes to this part of the codebase, and so I'll happily let other folks make changes to that specific dropdown in a separate PR, if necessary. We can keep this PR focused on the Popover arrow fixes

ciampo committed:
Use SCSS variable instead of CSS custom property

ciampo commented:
All feedback should be addressed. I've also rebased and removed the temporary code for the Storybook example and the `ownerDocument` [fix](https://github.com/WordPress/gutenberg/pull/42874#discussion_r935239671) (which I'll open separate PRs for).

If no one has any objections, I'm planning on merging this PR later today.
